pkg/tool/list.go

```diff
@@ -0,0 +1,310 @@
+		Version:          v.Version,
+		Cli:              cliPath,
+		Hub:              v.Hub,
+		ReleaseURLPrefix: v.ReleaseURLPrefix,
+	}
+}
+
+func (l *listOptions) Run(out io.Writer) error {
+	cs, err := yaml.Marshal(l.components)
+	if err != nil {
+		return fmt.Errorf("conversion failed. %v", err)
+	}
+
+	w := &toolListWriter{
+		cs: cs,
+	}
+
+	for _, v := range l.components {
+		w.entry = append(w.entry, v)
+	}
+	// sort by component name
+	sort.Slice(w.entry, func(i, j int) bool {
+		return w.entry[i].Name < w.entry[j].Name
+	})
+
+	return w.PrintObj(out, l.output)
+}
+
+type toolListWriter struct {
+	entry []componentElement
+	cs    []byte
+}
+
+func (t *toolListWriter) writeTable(out io.Writer) error {
+	table := uitable.New()
+
+	table.AddRow("NAME", "VERSION", "STATUS")
+	for _, te := range t.entry {
+		table.AddRow(te.Name, te.Version, te.Status)
+	}
+
+	return t.encodeTable(out, table)
+}
+
+func (t *toolListWriter) writeTableWIDE(out io.Writer) error {
+	table := uitable.New()
+
+	table.AddRow("NAME", "VERSION", "STATUS", "CLI", "HUB", "RELEASE-URL-PREFIX")
+	for _, te := range t.entry {
+		table.AddRow(te.Name, te.Version, te.Status, te.Cli, te.Hub, te.ReleaseURLPrefix)
+	}
+
+	return t.encodeTable(out, table)
+}
+
+func (t *toolListWriter) encodeTable(out io.Writer, table *uitable.Table) error {
+	raw := table.Bytes()
+	raw = append(raw, []byte("\n")...)
+	_, err := out.Write(raw)
+	if err != nil {
+		return fmt.Errorf("unable to write table output. %v", err)
+	}
+	return nil
+}
+
+func (t *toolListWriter) writeJSON(out io.Writer) error {
+	js, err := yaml.YAMLToJSON(t.cs)
+	if err != nil {
+		return fmt.Errorf("yaml cannot convert json. %v", err)
+	}
+
+	var prettyJSON bytes.Buffer
+	if err := json.Indent(&prettyJSON, js, "", "\t"); err != nil {
+		return fmt.Errorf("can't format json. %v", err)
+	}
+
+	if _, err := prettyJSON.WriteTo(out); err != nil {
+		return fmt.Errorf("unable to write json output. %v", err)
+	}
+	return nil
+}
+
+func (t *toolListWriter) writeYAML(out io.Writer) error {
+	if _, err := out.Write(t.cs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *toolListWriter) PrintObj(out io.Writer, format string) error {
+	switch strings.ToLower(format) {
+	case Table.String():
+		return t.writeTable(out)
+	case TableWIDE.String():
+		return t.writeTableWIDE(out)
+	case JSON.String():
+		return t.writeJSON(out)
+	case YAML.String():
+		return t.writeYAML(out)
+	}
+
+	return fmt.Errorf("invalid format type")
+}
+
+// cliToolName returns the Cli name of the component
+// When the newly added component has cli, need to add cli name.
+func cliToolName(componentName string) string {
+	switch strings.ToLower(componentName) {
+	case "istio":
+		return "istioctl"
+	case "karmada":
+		return "kubectl-karmada"
+	case "submariner":
+		return "subctl"
+	case "kubeedge":
+		return "keadm"
+	case "argocd":
+		return "argocd"
+	default:
+		return ""
+	}
+}
+
+func isReady(path string) (string, string) {
```

Comment:
prefer to return a boolean

Reply:
thanks for your review.
The name is confusing, i change the method name .